File: baekjoon/src/greedy/No13305.java
package greedy;

import java.io.*;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/13305
 */
public class No13305 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void sortCosts(int[] costs) {
        int min = Integer.MAX_VALUE;
        for (int i = 0; i < costs.length; i++) {
            if (min > costs[i])
                min = costs[i];
            else
                costs[i] = min;
        }
    }

    public static void main(String[] args) throws IOException {
        int numberOfInput = Integer.parseInt(reader.readLine());

        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        int[] lengths = new int[numberOfInput - 1];
        for (int i = 0; i < numberOfInput - 1; i++)
            lengths[i] = Integer.parseInt(tokenizer.nextToken());

        tokenizer = new StringTokenizer(reader.readLine());
        int[] costs = new int[numberOfInput];
        for (int i = 0; i < numberOfInput; i++)
            costs[i] = Integer.parseInt(tokenizer.nextToken());

        sortCosts(costs);

        long cost = 0;
        for (int i = 0; i < numberOfInput - 1; i++)
            cost += (long) costs[i] * lengths[i];

        writer.write(String.valueOf(cost));
        writer.flush();
    }
}

File: baekjoon/src/tags/greedy/No13305.kt
package tags.greedy

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

/**
 * https://www.acmicpc.net/problem/13305
 */
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))

    val cities = reader.readLine().toInt()
    val distances = reader.readLine().split(' ').map { it.toInt() }
    val costs = reader.readLine().split(' ').map { it.toInt() }

    var min = Integer.MAX_VALUE
    var cost = 0L
    for (i in 0 until cities - 1) {
        min = min.coerceAtMost(costs[i])
        cost += min.toLong() * distances[i]
    }

    writer.write(cost.toString())
    writer.flush()
}
